### DISCUSSION TITLE
Custom build of ipxe.efi fail to work in Hyper-V virtual machine

### DISCUSSION THREAD

**Discussion by ccy** [ipxe/ipxe#427] [General]
Summary of trial:

| No | Platform | Release | Commit | Intel NUC | Hyper-V |
|:-:|-|-|-|:-:|:-:|
| 1 | `ipxe.efi` | 1.0.0 | 53e9fb50 | ✔️ | ✔️ |
| 2 | `ipxe.efi` |  latest | http://boot.ipxe.org/ipxe.efi |  ✔️ | ✔️ |
| 3 | `snponly.efi` | 1.21 | e09e1142 | ✔️ | ✔️ |
| 4 | `ipxe.efi` | 1.21 | e09e1142 | ✔️ | ❌|

<ins>**Try 1: success**</ins>

I built and run `ipxe.efi` release 1.0.0+(commit 53e9fb50) with embedded script:

    make -j4 bin-x86_64-efi/ipxe.efi EMBED=boot.ipxe

on these 2 machines:

1. Intel NUC 10i7FNH. The NIC is **Intel I219-V (10)**.
2. A virtual machine in Hyper-V.

The `boot.ipxe` is a simple shell script only:

```
#!ipxe
shell
```

Upload `ipxe.efi` to TFTP server, both machines can download and boot the `ipxe.efi` without any issues.

<ins>**Try 2: success**</ins>

I also try the latest `http://boot.ipxe.org/ipxe.efi` (build 1.21) and it works too.

<ins>**Try 3: Success**</ins>

I then try custom build `snponly.efi` and it works for both machines.

<ins>**Try 4: fail on Hyper-V**</ins>

Next, I try to build `ipxe.efi` from master branch (commit e09e1142).  Upload the `ipxe.efi` to same TFTP server.  The new `ipxe.efi` works on Intel NUC but fail to work on the Hyper-V machine stop at this point:

    Downloading NBP file...

The TFTP server log show:

```
Aug  5 09:45:51 nas tftpd[73192]: Filename: 'ipxe/ipxe-x86_64.efi'
Aug  5 09:45:51 nas tftpd[73192]: Mode: 'octet'
Aug  5 09:45:51 nas tftpd[73192]: Got ERROR packet: User aborted the transfer
Aug  5 09:45:51 nas tftpd[73194]: Filename: 'ipxe/ipxe-x86_64.efi'
Aug  5 09:45:51 nas tftpd[73194]: Mode: 'octet'
Aug  5 09:45:51 nas tftpd[73194]: 192.168.0.233: read request for //ipxe/ipxe-x86_64.efi: success
Aug  5 09:45:56 nas tftpd[73194]: Timeout #0 on ACK 17
Aug  5 09:46:01 nas tftpd[73194]: Timeout #1 on ACK 17
Aug  5 09:46:06 nas tftpd[73194]: Timeout #2 on ACK 17
Aug  5 09:46:11 nas tftpd[73194]: Timeout #3 on ACK 17
Aug  5 09:46:16 nas tftpd[73194]: Timeout #4 on ACK 17
Aug  5 09:46:21 nas tftpd[73194]: Timeout #5 send ACK 17 giving up
```
I suspect it may be due to the block size issue of TFTP server. A more detail log shows that the block size is `1482`:

```
Aug  4 15:29:27 nas tftpd[38818]: Full access allowedin /etc/hosts.allow
Aug  4 15:29:27 nas tftpd[38818]: Filename: 'ipxe/ipxe-x86_64.efi'
Aug  4 15:29:27 nas tftpd[38818]: Mode: 'octet'
Aug  4 15:29:27 nas tftpd[38818]: option: 'tsize' value: '0'
Aug  4 15:29:27 nas tftpd[38818]: option: 'blksize' value: '1482'
Aug  4 15:29:27 nas tftpd[38818]: Setting blksize to '1482'
Aug  4 15:29:27 nas tftpd[38818]: Sending OACK
Aug  4 15:29:27 nas tftpd[38818]: Waiting 5 seconds for packet
Aug  4 15:29:27 nas tftpd[38818]: Got ERROR packet: User aborted the transfer
Aug  4 15:29:27 nas tftpd[38818]: Aborting: Error packet
Aug  4 15:29:27 nas tftpd[38820]: Full access allowedin /etc/hosts.allow
Aug  4 15:29:27 nas tftpd[38820]: Filename: 'ipxe/ipxe-x86_64.efi'
Aug  4 15:29:27 nas tftpd[38820]: Mode: 'octet'
Aug  4 15:29:27 nas tftpd[38820]: option: 'blksize' value: '1482'
Aug  4 15:29:27 nas tftpd[38820]: Setting blksize to '1482'
Aug  4 15:29:27 nas tftpd[38820]: Sending OACK
Aug  4 15:29:27 nas tftpd[38820]: Waiting 5 seconds for packet
Aug  4 15:29:27 nas tftpd[38820]: Received 4 bytes in a ACK packet
Aug  4 15:29:27 nas tftpd[38820]: 192.168.0.233: read request for //ipxe/ipxe-x86_64.efi: success
```

I trace the Intel NUC machine and it use blocksize of `1462`.

If blocksize is the issue, it should have failed for all the trials on Hyper-V.

If blocksize is not the issue, then why **trial 2** works but **trial 4** fail?  Is there special build option for the binary efi in http://boot.ipxe.org?

**Comment by NiKiZe** [comment]
You have one local build working, and latest failing correct? if so try [git bisect](https://ipxe.org/howto/bisect) to locate where this issue started.
Also check which version of gcc you have, might be a different version for boot.ipxe.org which makes a difference.

**Reply by ccy** [reply]
I compile iPXE on Windows WSL2 Debian instance.  The compiler is `gcc (Debian 8.3.0-6) 8.3.0`.  Using same source code and branch.  `snponly.efi` works but `ipxe.efi` fail to load on **Hyper-V VM's PXE**.

As the iPXE binary haven't fully download onto the machine for execution yet, I don't think it has any relation with the iPXE.

The file size of custom build `snponly.efi` is `186,880` bytes (works).
**The file size of custom build `ipxe.efi` is `1,007,104` bytes (fail).**
The file size of `http://boot.ipxe.org/ipxe.efi` is `1,009,152` bytes (works).

It is very strange why the custom build ipxe.efi can't download successfully to Hyper-V VM.

**Reply by ccy** [reply]
I think I almost found the cause but I still don't understand why.

I am using a [FS-S3900](https://www.fs.com/products/72944.html) and a mikrotik network switch.

TFTP machine connect to switch FS switch and Hyper-V connect to mikrotik switch. Then I face the custom build `ipxe.efi` can't success download issue.

I replace the mikrotik network switch to other switch and the problem gone.

I suspect it might due to mikrotik network switch's MTU or jumbo frame setting.  Unfortunately, I couldn't find out the cause.
